docs: :memo: add conversion report section to design (#252)
dp-next/registers2parquet: vignettes/design.qmd

requirements, the public-facing interface, and some diagrams
highlighting the general flow of the main functions.

::: callout-note
Using R to read SAS can't guarantee perfect preservation of the SAS
values, since reading SAS files in R relies on

> [`haven`](https://haven.tidyverse.org/index.html), which is a

Reviewer: ```suggestion
[haven](https://haven.tidyverse.org/index.html), which is based on
```

They don't use backticks

> reverse-engineered effort to read the proprietary SAS file format.

Reviewer: ```suggestion
[ReadStat](https://github.com/WizardMac/ReadStat), a reverse-engineered effort to read the proprietary SAS file format.
```

> However, `haven` is a mature library and explicitly supports reading

Reviewer: ```suggestion
However, haven and the underlying ReadStat are mature packages and explicitly support reading
```

    path --> read_register --> output

```

## Conversion report

When `convert_file()` is called with `report = TRUE` (the default), a
conversion report is written to `output_dir`. The report compares each
SAS chunk with its corresponding Parquet file to check that the
conversion was successful.

> Columns that are identical in the SAS chunk and Parquet file are

Reviewer: How might you think this is compared? The data read into R from SAS and saved into Parquet will be the exact same as both are in the same source environment (R). Are you thinking maybe we load in the Parquet and compare read in SAS vs read in Parquet?

Author: @lwjohnst86 did you have some thoughts on writing to the same file with multiple workers in parallel?

Reviewer: As I mentioned in my other comments and via discord, the problem remains, that in both cases of reading from SAS or Parquet, we are reading into R. And types between SAS, Parquet, and R don't perfectly match. So we would always be comparing R types with R types.

    path --> read_register --> output

```

> ## Conversion report

Reviewer: Part of the purpose of a design is the *why*. Why do we want to add a report ability? What will it be used for? How will it be used? What is the need being fulfilled here? I'm a bit uncertain what that need is (at least based on what you've put here).

When I think of this need (at least for myself, if I think back to when I was processing stuff) is these things:

- Are the columns in each individual Parquet file of a given register the same type?
- Are the column names the same and are there the same number of columns?

These questions are because I want to make sure that when I read in the whole Parquet dataset for a register (with `unify_schemas`), that no columns are accidentally the wrong type or there's extra columns that are dropped and I don't even know about it. Mainly these are coming from my limited understanding of how exactly `unify_schemas` works. I *think* it's taking the first part file and applying that schema to all part files in the dataset.

When `convert_file()` is called with `report = TRUE` (the default), a
conversion report is written to `output_dir`. The report compares each
SAS chunk with its corresponding Parquet file to check that the
conversion was successful.

The Parquet file is read into R and compared column-by-column against
the SAS chunk. Identical columns are marked [OK], while differing

> columns are marked [DIFF] showin with a side-by-side comparison. Column

Reviewer: ```suggestion
columns are marked [DIFF] shown with a side-by-side comparison. Column
```

names are lowercased and the types are Arrow types, matching the Parquet
file.

Statistics depend on the column type:

- **Numeric** (`int32`, `float64`): min, max, mean, median, Q25, Q75, NA
  count
- **Datetime** (`timestamp`, `date32`): min, max, NA count
- **Categorical** (`large_utf8`, `bool`): NA count + top 5 value counts;

>   remaining values shown as `(+N more)`

Reviewer: Based on my comment above about the *need*/what this is solving, I'm not sure these are relevant/necessary. What's the actual use case here?

> Below is an example of how the report is structured:

Reviewer: Rather than output a file, output a "JSON style" output. Something like this could be the output:

``` r
list(
	input_path = "",
	output_path = "",
	column_info = tibble(
		name = ...,
		type = ...
	)
)
```

And actually, this could probably be achieved by a separate function that reads the individual Parquet `part-` files and calculates this, rather than done through `convert_file()`.

    path --> read_register --> output

```

## Conversion report

The purpose of the conversion report is to check that the conversion
went as expected (to the extent that's possible). It addresses two
checks:

1. Do the SAS file and the output Parquet file(s) have the same number

>    of rows?

Reviewer: Rather than have that part of the design of a report, we can have tests that confirm the size before and after. But this really shouldn't be something to check, because there's nothing we can do if **one of the read/write packages doesn't correctly save the data**. This is the really important part here. A report should give the user actionable items to resolve the issue. But if the rows are different, what can the user actually do about it? If they are different, that means the haven and arrow packages are flawed and that's way outside the scope of fastreg.

Reviewer: To add to this, I think that many of these things can also be tested here to confirm what would actually happen. E.g. like if have chunks of 100, and the 101th chunk has a schema that's different from the first chunk, we will know what would actually happen here.

Author: That is a good point; that a row count mismatch isn't actionable in the sense that the user can't fix a bug in haven or arrow. But, I do think we can think of this not as a check that prompts the user to *do* something, but as a kind of conversion "audit trail". When I convert the SAS fils, I would like something concrete to point to, not as a debugging step, but as information to say "for this specific register, the conversion resulted in the same number of rows." The tests check the pipeline works in general; they don't document what happened in a particular conversion. Does that make sense?

2. Are the column names, number of columns, and data types the same
   across all Parquet files within a register (e.g., the files created
   from `bef2020.sas7bdat` and `bef2021.sas7bdat`)?

The report is produced by three functions with distinct
responsibilities:

- `convert_file()`: Converts a single SAS file and returns the input

>   path, the SAS row count, and the paths of the Parquet parts written.

Reviewer: This isn't necessary based on my comment above. Also, it isn't good to complicate the interface in this way, at least in regards to the row count. But, I like where your thinking is going with outputting logging information from this function 👍

> - `collect_parquet_info()`: Reads the Parquet parts for a single file

Reviewer: ```suggestion
- `get_parquet_info()`: Reads the Parquet parts for a single file
```

Rather than introduce another verb.

Reviewer: You could also include the function signature and output for this.

>   and returns the total row count, column names and data types.

Reviewer: ```suggestion
  and returns the column names and data types.
```

- `create_report()`: Assembles the report by taking the outputs of the
  above two functions for all files in a register and returns a

>   structured list with the row count comparison and column information.

Reviewer: ```suggestion
- `create_report()`: Assembles the report by taking the outputs of the
  above two functions for all files in a register and returns a
  structured list with a comparison and basic check of the files.
```

In the targets pipeline, `convert_file()` and `collect_parquet_info()`
run per SAS file. A single downstream `create_report()` target assembles
the final report once all conversions are complete.

> ### Check 1: Row count comparison

Reviewer: See my comment above about row counts.

`convert_file()` tracks the number of rows read from the SAS file via
the loop counter and returns it. `collect_parquet_info()` reads the
written Parquet parts and returns their total row count.
`create_report()` compares the two and records whether they match.

A row count match does not guarantee that the values themselves are
preserved correctly — it only confirms that no rows were lost during
conversion. Its primary value is as an auditable record: if the counts
match, there is a concrete data point to refer to when checking that a
conversion completed as expected.

::: callout-warning
This check cannot guarantee perfect preservation of the SAS values,
since reading SAS files in R relies on `haven`, which is a
reverse-engineered effort to read the proprietary SAS file format.

> :::

Author: This check adds the complexity of having to change the return of `convert_file()` to include the SAS row count, and I expect you to push back on this still (and maybe rightfully so). I would just really like to have something to point to that compares the written Parquet file(s) to the originally read in SAS file. Looking forward to your comments :grin:

> :::

Reviewer: I think this comment could even be right up near the top this page, to highlight that if you use R to read in SAS, you can never truly guarantee that it is being correctly read in. But that so many people depend on and use packages like haven, that if there was a true issue, it probably would have been found or will eventually be found.

Author: I've moved it up 👍

### Check 2: Columns and data types across Parquet files within a register

Since we can't be sure that the SAS files within the same register
contain exactly the same columns and data types, the conversion report
helps identify any differences between these files.

::: callout-note
Discrepancies (different columns or incompatible data types) between
files within the same register do not stop the conversion, but will be
noted in the report.
:::

`collect_parquet_info()` reads each Parquet part and collects column
names and types. `create_report()` compares these across all files and

> reports a conclusive result for each register:

Reviewer: ```suggestion
reports on the findings of the checks for the register:
```

    %% Edges
    name --> read_register --> output
```

> ## Conversion report

Reviewer: ```suggestion
## Conversion log
```

> The purpose of the conversion report is to check that the conversion

Reviewer: ```suggestion
The purpose of the conversion log is to describe the details of the conversion
```

> went as expected so the partitioned Parquet register(s) can be read into

Reviewer: ```suggestion
to provide an audit trail.
```

> R collectively afterwards. Since we can't be sure that the SAS files

Reviewer: ```suggestion
Since we can't be sure that the SAS files
```

within the same register contain exactly the same columns and data

> types, the conversion report helps identify any differences between

Reviewer: ```suggestion
types, the conversion log helps identify any differences between
```

these files.

::: callout-note
Discrepancies (different columns or incompatible data types) between

> files within the same register do not stop the conversion, but will be

Author: ```suggestion
files within the same register do not stop the conversion, but it will be
```

> noted in the report.

Reviewer: ```suggestion
files within the same register do not stop the conversion, but it will be included in the log.
```

:::

> `convert_file()` returns a metadata tibble with one row per written

Reviewer: ```suggestion
`convert()` returns a metadata tibble with one row per written
```

> chunk. This can be queried with `dplyr` directly or rendered into a

Reviewer: ```suggestion
chunk. This can be queried with dplyr directly or rendered into a
```

Quarto report.

> ### Return value of `convert_file()`

Reviewer: ```suggestion
### Return value of `convert()`
```

> `convert_file()` returns a tibble with one row per written chunk:

Reviewer: ```suggestion
`convert()` returns a tibble with one row per written chunk:
```

| Column         | Description                                  |
|----------------|----------------------------------------------|
| `input_path`   | Path to the source SAS file                  |
| `output_path`  | Path to the written Parquet part file        |
| `row_count`    | Number of rows in the chunk                  |

> | `column_count` | Number of columns in the chunk               |

Author: I guess this could be derived from `columns`, so we could remove this.

> | `column_count` | Number of columns in the chunk               |

Reviewer: ```suggestion
```

| `columns`      | Nested tibble with column `name` and `type`  |

The information is derived from the chunk already in memory, not by
reading the Parquet file back.

```r
# Before repeat loop.

> chunk_info_list <- list()

Reviewer: ```suggestion
chunk_info <- tibble::tribble(
  ~input_path, ~output_path, ~row_count, ~columns
)
```

This allows creating an empty tibble with the correct columns

# Inside the repeat loop, after writing.
chunk_info_list <- c(chunk_info_list, list(tibble::tibble(
  input_path   = path,
  output_path  = fs::path(file_path),
  row_count    = nrow(chunk),
  column_count = ncol(chunk),
  columns      = list(tibble::tibble(
    name = colnames(chunk),
    type = purrr::map_chr(chunk, class)
  ))

> )))

Reviewer: ```suggestion
chunk_info <- dplyr::bind_rows(
  chunk_info, 
  tibble::tibble(
    input_path   = path,
    output_path  = fs::path(file_path),
    row_count    = nrow(chunk),
    columns      = tibble::tibble(
      name = colnames(chunk),
      type = purrr::map_chr(chunk, class)
    )
  )
)
```

Doesn't need to be a list

# After the loop, bind all chunk tibbles.

> dplyr::bind_rows(chunk_info_list)

Reviewer: ```suggestion
```

Author: ```suggestion

# After the loop, return the collected information.
chunk_info
```

```

### Quarto report template

> `use_fastreg_template()` copies both `_targets.R` and

Author: ```suggestion
`use_template()` copies both `_targets.R` and
```

> `conversion_report.qmd` into the current working directory. The Quarto

Author: ```suggestion
`conversion_log.qmd` into the current working directory. The Quarto
```

doc reads `chunk_info` via `targets::tar_read()` and produces an HTML or

> PDF report for review.

Author: ```suggestion
PDF report for review. The default is PDF, but it can easily be changed in the Quarto file.
```

> <!-- TODO: What should the default format be? -->

Reviewer: Maybe PDF via typst, since that's easier to view on DST. Outside of DST, I would say HTML, but DST is special.

```r
chunk_info <- targets::tar_read(chunk_info)

# Nice overview of the info + schema comparison within registers.
...
```

The report is added to the targets pipeline as a last target:

```r

> tar_target(

Reviewer: ```suggestion
tar_quarto(
```

This relies on tarchetypes package
https://docs.ropensci.org/tarchetypes/reference/tar_quarto.html

  name = report,

>   command = quarto::quarto_render("conversion_report.qmd"),

Reviewer: ```suggestion
  path = "conversion_report.qmd"
```

>   deployment = "main"

Reviewer: ```suggestion
```

files within the same register do not stop the conversion, but it will
be included in the log.
:::

`convert()` returns a metadata tibble with one row per written chunk.
This can be queried with dplyr directly or rendered into a Quarto log.

### Return value of `convert()`

`convert()` returns a tibble with one row per written chunk:

| Column         | Description                                  |
|----------------|----------------------------------------------|
| `input_path`   | Path to the source SAS file                  |
| `output_path`  | Path to the written Parquet part file        |
| `row_count`    | Number of rows in the chunk                  |
| `columns`      | Nested tibble with column `name` and `type`  |

The information is derived from the chunk already in memory, not by
reading the Parquet file back.

```r
# Before repeat loop.
chunk_info <- tibble::tribble(
  ~input_path, ~output_path, ~row_count, ~columns
)

# Inside the repeat loop, after writing.
chunk_info <- dplyr::bind_rows(
  chunk_info,
  tibble::tibble(
    input_path   = path,
    output_path  = fs::path(file_path),
    row_count    = nrow(chunk),

>     columns      = tibble::tibble(
>       name = colnames(chunk),
>       type = purrr::map_chr(chunk, class)
>     )

Author: ```suggestion
    columns      = list(tibble::tibble(
      name = colnames(chunk),
      type = purrr::map_chr(chunk, class)
    ))
```
Oh, this needs to be a list, otherwise we get multiple rows for the same chunk.

Reviewer: ? I'm not sure I understand what you mean. It couldn't be multiple rows per chunk if the column is a tibble.

Reviewer: When you use `tibble()` in a column within a `tibble()`, it creates a nested tibble. So it would be one row per chunk.
